[Postgres RLS] Issue with keyword_view

Seen on running perf test with postgres rls:

{"timestamp":"2026-04-15T07:10:43.992Z","level":"ERROR","thread":"reactor-tcp-epoll-4","mdc":{"eventId":"9bfde370-de31-4fe0-bb39-cd0450e2c2d0","eventClass":"org.apache.james.mailbox.events.MailboxEvents.FlagsUpdated","user":"firstname2777.surname2777@upn.integration-open-paas.org","group":"com.linagora.tmail.james.jmap.event.KeywordEmailQueryViewListener$KeywordEmailQueryViewListenerGroup"},"logger":"org.apache.james.events.GroupConsumerRetry","message":"Exception happens when handling event after 3 retries","context":"default","exception":"org.jooq.exception.DataAccessException: SQL [delete from keyword_view\nwhere (\n  username = $1\n  and keyword = $2\n  and received_at = cast($3 as timestamp with time zone)\n  and message_id = cast($4 as uuid)\n)]; unrecognized configuration parameter \"app.current_domain\"\n\tat org.jooq.impl.Tools.translate(Tools.java:3681)\n\tat org.jooq.impl.Tools.translate(Tools.java:3644)\n\tat org.jooq.impl.Tools.translate(Tools.java:3628)\n\tat org.jooq.impl.R2DBC$Forwarding.lambda$onError$0(R2DBC.java:288)\n\tat org.jooq.impl.R2DBC$AbstractNonBlockingSubscription.lambda$cancel0$5(R2DBC.java:773)\n\tat java.base/java.util.concurrent.atomic.AtomicReference.updateAndGet(Unknown Source)\n\tat org.jooq.impl.R2DBC$AbstractNonBlockingSubscription.cancel0(R2DBC.java:761)\n\tat org.jooq.impl.R2DBC$AbstractSubscription.complete(R2DBC.java:226)\n\tat org.jooq.impl.R2DBC$AbstractResultSubscriber.complete(R2DBC.java:349)\n\tat org.jooq.impl.R2DBC$Forwarding.complete(R2DBC.java:301)\n\tat org.jooq.impl.R2DBC$Forwarding.onError(R2DBC.java:288)\n\tat org.jooq.impl.DefaultSubscriberProvider$1.onError(DefaultSubscriberProvider.java:86)\n\tat org.jooq.impl.DefaultSubscriberProvider$1.onError(DefaultSubscriberProvider.java:86)\n\tat reactor.core.publisher.StrictSubscriber.onError(StrictSubscriber.java:106)\n\tat reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onError(FluxHandleFuseable.java:229)\n\tat reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onError(MonoCollectList.java:108)\n\tat reactor.core.publisher.FluxHandle$HandleSubscriber.onError(FluxHandle.java:213)\n\tat reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onError(MonoFlatMapMany.java:256)\n\tat reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:201)\n\tat reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onNext(FluxFilterFuseable.java:337)\n\tat reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:854)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:854)\n\tat io.r2dbc.postgresql.util.FluxDiscardOnCancel$FluxDiscardOnCancelSubscriber.onNext(FluxDiscardOnCancel.java:91)\n\tat reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:113)\n\tat reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:129)\n\tat reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:880)\n\tat reactor.core.publisher.FluxCreate$BufferAsyncSink.next(FluxCreate.java:805)\n\tat reactor.core.publisher.FluxCreate$SerializedFluxSink.next(FluxCreate.java:163)\n\tat io.r2dbc.postgresql.client.ReactorNettyClient$Conversation.emit(ReactorNettyClient.java:696)\n\tat io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.emit(ReactorNettyClient.java:948)\n\tat io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:822)\n\tat io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:728)\n\tat reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:129)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:854)\n\tat reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)\n\tat reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)\n\tat reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:292)\n\tat reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:401)\n\tat reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:435)\n\tat reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:115)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)\n\tat io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)\n\tat io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)\n\tat io.r2dbc.postgresql.client.SSLSessionHandlerAdapter.channelRead(SSLSessionHandlerAdapter.java:85)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)\n\tat io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1519)\n\tat io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1377)\n\tat io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1428)\n\tat io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530)\n\tat io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)\n\tat io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)\n\tat io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)\n\tat io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:501)\n\tat io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:399)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: io.r2dbc.postgresql.ExceptionFactory$PostgresqlBadGrammarException: unrecognized configuration parameter \"app.current_domain\"\n\tat io.r2dbc.postgresql.ExceptionFactory.createException(ExceptionFactory.java:96)\n\tat io.r2dbc.postgresql.ExceptionFactory.createException(ExceptionFactory.java:65)\n\tat io.r2dbc.postgresql.ExceptionFactory.handleErrorResponse(ExceptionFactory.java:132)\n\tat reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:179)\n\t... 55 common frames omitted\n"}

--- a/tmail-backend/jmap/extensions-postgres/src/main/java/com/linagora/tmail/james/jmap/projections/PostgresKeywordEmailQueryView.java
+++ b/tmail-backend/jmap/extensions-postgres/src/main/java/com/linagora/tmail/james/jmap/projections/PostgresKeywordEmailQueryView.java
@@ -33,7 +33,6 @@ import java.util.Optional;
 import java.util.function.Function;
 
 import jakarta.inject.Inject;
-import jakarta.inject.Named;
 
 import org.apache.james.backends.postgres.utils.PostgresExecutor;
 import org.apache.james.core.Username;
@@ -53,16 +52,16 @@ import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
 public class PostgresKeywordEmailQueryView implements KeywordEmailQueryView {
-    private final PostgresExecutor postgresExecutor;
+    private final PostgresExecutor.Factory executorFactory;
 
     @Inject
-    public PostgresKeywordEmailQueryView(@Named(PostgresExecutor.BY_PASS_RLS_INJECT) PostgresExecutor postgresExecutor) {
-        this.postgresExecutor = postgresExecutor;
+    public PostgresKeywordEmailQueryView(PostgresExecutor.Factory executorFactory) {
+        this.executorFactory = executorFactory;
     }
 
     @Override
     public Mono<Void> save(Username username, Keyword keyword, Instant receivedAt, MessageId messageId, ThreadId threadId) {
-        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME)
+        return executorFactory.create(username.getDomainPart()).executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME)
             .set(USERNAME, username.asString())
             .set(KEYWORD, keyword.flagName())
             .set(RECEIVED_AT, OffsetDateTime.ofInstant(receivedAt, ZoneOffset.UTC))
@@ -74,7 +73,7 @@ public class PostgresKeywordEmailQueryView implements KeywordEmailQueryView {
 
     @Override
     public Mono<Void> delete(Username username, Keyword keyword, Instant receivedAt, MessageId messageId) {
-        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+        return executorFactory.create(username.getDomainPart()).executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
             .where(USERNAME.eq(username.asString()))
             .and(KEYWORD.eq(keyword.flagName()))
             .and(RECEIVED_AT.eq(OffsetDateTime.ofInstant(receivedAt, ZoneOffset.UTC)))
@@ -83,14 +82,15 @@ public class PostgresKeywordEmailQueryView implements KeywordEmailQueryView {
 
     @Override
     public Flux<MessageId> listMessagesByKeyword(Username username, Keyword keyword, Options options) {
+        PostgresExecutor executor = executorFactory.create(username.getDomainPart());
         return EmailQueryViewUtils.QueryViewExtender.of(options.limit(), options.collapseThread())
-            .resolve(backendFetchLimit -> postgresExecutor.executeRows(dslContext -> Flux.from(buildSelectStatement(dslContext, username, keyword, options, backendFetchLimit)))
+            .resolve(backendFetchLimit -> executor.executeRows(dslContext -> Flux.from(buildSelectStatement(dslContext, username, keyword, options, backendFetchLimit)))
                 .map(asEmailEntry()));
     }
 
     @Override
     public Mono<Void> clearAll() {
-        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.truncate(TABLE_NAME)));
+        return executorFactory.create().executeVoid(dslContext -> Mono.from(dslContext.truncate(TABLE_NAME)));
     }
 
     private SelectLimitPercentStep buildSelectStatement(DSLContext dslContext, Username username, Keyword keyword, Options options, Limit backendFetchLimit) {

--- a/tmail-backend/jmap/extensions-postgres/src/test/java/com/linagora/tmail/james/jmap/projections/PostgresKeywordEmailQueryViewTest.java
+++ b/tmail-backend/jmap/extensions-postgres/src/test/java/com/linagora/tmail/james/jmap/projections/PostgresKeywordEmailQueryViewTest.java
@@ -39,7 +39,7 @@ public class PostgresKeywordEmailQueryViewTest implements KeywordEmailQueryViewC
 
     @Override
     public KeywordEmailQueryView testee() {
-        return new PostgresKeywordEmailQueryView(postgresExtension.getDefaultPostgresExecutor());
+        return new PostgresKeywordEmailQueryView(postgresExtension.getExecutorFactory());
     }
 
     @Override

